### ISSUE TITLE
memoryview, python-side allocation

### ISSUE THREAD

**Issue by elbaro** [PyO3/pyo3#617]
- Feature request: PyMemoryView
- Feature request: PyByteArray::new_with_size(num_bytes)
- Question: Can we allocate some buffer in python-side?

Consider this example:
```
let vec = vec![0;length];
// write on vec
// ...
return PyBytes::new(&vec[..]).to_object(py);
```

PyBytes::new is very slow because it copies. A memoryview on Rust slice works. but it is hard to manage the lifetime. If we can allocate on the python side from the beginning,  the copy is not necessary. Is this possible?

```
let buf = some_allocation(length);
let bytes = buf.as_mut_bytes();
// write directly on bytes
// ...
return bytes;
```

**Comment by kngwyu** [comment]
Thank you for letting us know that.
[rust-numpy](https://github.com/rust-numpy/rust-numpy) has this kind of feature, so maybe we can implement this for `PyBuffer`.
But, since the current `PyBuffer` design is not good and complicated, it needs some more design decisions.

**Comment by davidhewitt** [comment]
I think we might be able to do something here with a closure like `PyBytes::new_with(py, 1024, |b: &mut [u8]| { ... }`.

My thinking is that perhaps during a controlled initialization in this fashion, nothing else can yet access the bytes object so it can be mutated safely.

If anyone strongly wants this feature, please open a PR and we can flesh it out.

**Comment by davidhewitt** [comment]
Closed in #1074
